go/consensus/tendermint/api/app.go

```diff
@@ -1,15 +1,51 @@
 package api
 
 import (
+	"errors"
+
 	tmabcitypes "github.com/tendermint/tendermint/abci/types"
 
 	"github.com/oasisprotocol/oasis-core/go/consensus/api/transaction"
 	genesis "github.com/oasisprotocol/oasis-core/go/genesis/api"
 )
 
+// ErrNoSubscribers is the error returned when publishing a message that noone is subscribed to.
+var ErrNoSubscribers = errors.New("no subscribers to given message kind")
+
+// MessageSubscriber is a message subscriber interface.
+type MessageSubscriber interface {
+	// ExecuteMessage executes a given message.
+	ExecuteMessage(ctx *Context, kind, msg interface{}) error
+}
+
+// MessageDispatcher is a message dispatcher interface.
+type MessageDispatcher interface {
+	// Subscribe subscribes a given message subscriber to messages of a specific kind.
+	Subscribe(kind interface{}, ms MessageSubscriber)
```

Comment:
Is it expected that message kinds will get more complex in future? 
If not, it might be worth considering a single type for all message kinds (instead of having per module types all defined the same way). Would also make the type signatures here more descriptive as `kinds` could be a concrete type and not `interface{}`

Reply:
The interface is not there to allow more complex kinds but rather to allow each module to define its own kinds in its respective `api` packages without needing a central place to list all kinds.

Reply:
Yeah that probably is indeed nicer :+1: